finetune/eval/method.py

```diff
@@ -250,6 +257,77 @@ def compute_multiple_choice_deviation(
     )
 
 
+def compute_if_eval(
+    model: PreTrainedModel,
+    tokenizer: transformers.PreTrainedTokenizer,
+    sequence_length: int,
+    batches: typing.List[typing.Tuple[torch.Tensor, typing.List[IFEvalRule]]],
+    device: str,
+) -> float:
+    """Computes the IFEval score for a given model on provided batches.
+
+    Args:
+        model (PreTrainedModel): The model for which losses are to be computed.
+        tokenizer (transformers.PreTrainedTokenizer): Tokenizer to tokenize the output with before returning.
+        sequence_length (int): The maximum length of the generated sequences.
+        batches (list): A list of batches containing prompts and rules.
+        device (str): The device to use for computation (e.g., 'cpu', 'gpu')."""
+    scores = []
+    duplicate_count = 0
+
+    # TODO: Figure out the right config to use.
+    generation_config = GenerationConfig(
+        max_length=sequence_length,
+        do_sample=False,
+        repetition_penalty=1.1,
+        eos_token_id=tokenizer.eos_token_id,
+        pad_token_id=tokenizer.eos_token_id,
+    )
+
+    for (
+        inputs,
+        rules,
+    ) in batches:
+        try:
+            responses = []
+            for input in inputs:
+                responses.append(
+                    generate_output(
+                        model=model,
+                        input_ids=input,
+                        generation_config=generation_config,
+                        device=device,
+                        tokenizer=tokenizer,
+                    )
+                )
+
+            # Check for overlap in the response
+            if len(responses) != 2:
+                raise ValueError(f"Expected 2 responses, got {len(responses)}")
+            if compute_similarity_score(responses[0], responses[1]) > 0.6:
+                duplicate_count += 1
+
+            for response in responses:
+                for rule in rules:
+                    if rule.matches(response):
+                        scores.append(0)
+                    else:
+                        scores.append(1)
+        except Exception as e:
+            bt.logging.error(
+                f"Exception occurred in multiple choice deviation computation: {e}"
+            )
+            traceback.print_exc()
+            scores.append(1)  # Use 1 to indicate failure
```

Comment:
Technically we should append 1 per rule here I think. To match if they got every rule wrong.